packages/rs-drive-abci/src/logging.rs

```diff
@@ -0,0 +1,627 @@
+/// See https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html#formatters
+#[derive(Debug, Serialize, Deserialize, Default, Clone, Copy)]
+pub enum LogFormat {
+    /// Default, human-readable, single-line logs
+    #[default]
+    Full,
+    ///  A variant of the default formatter, optimized for short line lengths
+    Compact,
+    /// Pretty, multi-line logs, optimized for human readability
+    Pretty,
+    /// Outputs newline-delimited JSON logs, for machine processing
+    Json,
+}
+
+/// Configuration of log destinations.
+///
+/// Logs can be sent to multiple destinations. Configuration of each of them is prefixed with `ABCI_LOG_<key>_`,
+/// where `<key>` is some arbitrary alphanumeric name of log configuaration.
+///
+/// Key must match pattern `[A-Za-z0-9]+`.
+///
+/// Note that parsing of LogConfigs is implemented in [PlatformConfig::from_env()] due to limitations of [envy] crate.
+///
+/// ## Example
+///
+/// ```bash
+/// # First logger, logging to stderr on verbosity level 5
+/// ABCI_LOG_StdErr_DESTINATION=stderr
+/// ABCI_LOG_StdErr_VERBOSITY=6
+///
+/// # Second logger, logging to stdout on verbosity level 1
+/// ABCI_LOG_StdOut_DESTINATION=stdout
+/// ABCI_LOG_StdOut_VERBOSITY=1
+/// ```
+///
+///
+/// [PlatformConfig::from_env()]: crate::config::PlatformConfig::from_env()
+pub type LogConfigs = HashMap<String, LogConfig>;
+
+impl FromEnv for LogConfigs {
+    /// create new object using values from environment variables
+    fn from_env() -> Result<Self, crate::error::Error>
+    where
+        Self: Sized,
+    {
+        let re: Regex = Regex::new(r"^ABCI_LOG_([0-9a-zA-Z]+)_DESTINATION$").unwrap();
+        let keys = std::env::vars().filter_map(|(key, _)| {
+            re.captures(&key)
+                .and_then(|capt| capt.get(1))
+                .map(|m| m.as_str().to_string())
+        });
+
+        let mut configs: HashMap<String, LogConfig> = HashMap::new();
+        for key in keys {
+            let config: LogConfig = envy::prefixed(format! {"ABCI_LOG_{}_", key.as_str()})
+                .from_env()
+                .map_err(crate::error::Error::from)?;
+
+            configs.insert(key.as_str().to_string(), config);
+        }
+
+        Ok(configs)
+    }
+}
+
+/// Errors returned by logging subsystem
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// File rotation error
+    #[error("file rotation: {0}")]
+    FileRotate(std::io::Error),
+
+    /// Invalid destination
+    #[error(
+        "invalid destination {0}: must be one of: stderr, stdout, or absolute path to a directory"
+    )]
+    InvalidDestination(String),
+
+    /// Log file path is invalid
+    #[error("log file path {0}: {1}")]
+    FilePath(PathBuf, String),
+
+    /// Duplicate config
+    #[error("duplicate log configuration name {0}")]
+    DuplicateConfigName(String),
+}
+
+/// Name of logging configuration
+pub type LoggerID = String;
+
+/// Add and manage logging methods
+pub struct LogController {
```

Comment:
It looks more like a mixed Builder pattern with the collection for logs. I would separate builder logic from the container, so builder will create a container.